lib/date_time/date_time_ntp.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <net/sntp.h>
+#include <net/socketutils.h>
+
+#if defined(CONFIG_LTE_LINK_CONTROL)
+#include <modem/lte_lc.h>
+#endif
+
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(date_time, CONFIG_DATE_TIME_LOG_LEVEL);
+
+#define UIO_IP      "ntp.uio.no"
+#define GOOGLE_IP_1 "time1.google.com"
+#define GOOGLE_IP_2 "time2.google.com"
+#define GOOGLE_IP_3 "time3.google.com"
+#define GOOGLE_IP_4 "time4.google.com"
+
+#define NTP_DEFAULT_PORT "123"
+
+struct ntp_servers {
+	const char *server_str;
+	struct sockaddr addr;
+	socklen_t addrlen;
```

Comment:
This reserves memory for IP addresses of all 5 NTP servers to remove the need to do a DNS query each time. First, this takes something like 100 bytes of RAM, although usually the device always uses the first server. Secondly, modem does some kind of caching for the addresses as well. And finally, at least Google offers load balancing for NTP and the same may be true for `ntp.uio.no` as well. To utilize load balancing, we need to do a DNS query.

Google uses load balancing with address `time.google.com`. This address is resolved to one of the Google NTP servers depending on the location and load. I also tried `ntp.uio.no` and that got resolved to the same IP as `ns1.uio.no`, and there's also `ns2.uio.no`. So, it might be that `ntp.uio.no` is also resolved to either one of those servers.

My question is, would it be better to modify the implementation so, that there are only two NTP server addresses (`ntp.uio.no` and `time.google.com`) and `getaddrinfo()` would be called every time? This would save some RAM, take advantage of the NTP server load balancing and simplify the implementation. It would increase the data consumption a bit, but not by much.

Reply:
That sounds like a valid improvement 👍